<div align="center">

# 🚢 Titanic Survival Classification  
### Predicting Passenger Survival (Binary Classification)

</div>

 ---
 
<div align="center">
    
## 📌Step 1: Define Objective

</div>

- **Goal:** Predict whether a passenger survived (1) or did not survive (0) the Titanic disaster.  
- **Problem Type:** Supervised **binary classification**.  
- **Target Column:** `Survived` (0 = No, 1 = Yes).  
- **Key Features:** `Pclass`, `Sex`, `Age`, `SibSp`, `Parch`, `Fare`, `Embarked` (and engineered features later).  

### Success Criteria / Evaluation Metrics
- **Primary Metrics:** Accuracy, ROC-AUC.  
- **Secondary Metrics:** Precision, Recall, F1-score, and Confusion Matrix to evaluate prediction quality.  

### Constraints & Considerations
- Limited dataset size → prefer simpler, interpretable models first.  
- Missing values in `Age`, `Cabin`, `Embarked` require imputation.  
- Possible class imbalance → evaluate using balanced metrics.

---

<div align="center">
    
## 📌Step 2: Load & Explore Data

</div>

## Objective
Load the Titanic dataset (`train.csv` and `test.csv`) and perform an initial exploration to understand its structure and contents.

### Key Tasks
- Load the training and test datasets using **pandas**.  
- View dataset **shape**, **columns**, and **data types**.  
- Identify **missing values** and **null percentages**.  
- Generate **summary statistics** for numerical features.  
- Examine the **distribution of the target variable** (`Survived`) to understand class balance.



In [1]:
# Step 2: Load & Explore Data

import pandas as pd

# Load datasets
train_df = pd.read_csv(r"C:\Users\AZAN LAPTOP STORE\Downloads\Portfolio\Projects\Titanic\train.csv")
test_df = pd.read_csv(r"C:\Users\AZAN LAPTOP STORE\Downloads\Portfolio\Projects\Titanic\test.csv")

# Basic info
print("Train shape:", train_df.shape)
print("Test shape:", test_df.shape)
print("\nTrain columns:\n", train_df.columns)

# Preview first few rows
print("\nTraining Data Preview:")
display(train_df.head())

# Check data types and missing values
print("\n--- Data Info ---")
train_df.info()

print("\n--- Missing Values ---")
print(train_df.isnull().sum())

# Summary statistics
print("\n--- Summary Statistics ---")
display(train_df.describe())

# Target distribution
print("\n--- Target Variable Distribution ---")
print(train_df['Survived'].value_counts(normalize=True) * 100)


Train shape: (891, 12)
Test shape: (418, 11)

Train columns:
 Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

Training Data Preview:


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S


--- Data Info ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB

--- Missing Values ---
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin      

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200


--- Target Variable Distribution ---
Survived
0    61.616162
1    38.383838
Name: proportion, dtype: float64


---

### Step 2 Insights – Load & Explore Data

- The **training dataset** contains **891 records** and **12 features**, while the **test dataset** contains **418 records** and **11 features** (missing the target column `Survived`).  
- Features include a mix of **categorical** (`Sex`, `Embarked`, `Cabin`, `Name`, `Ticket`) and **numerical** (`Age`, `Fare`, `SibSp`, `Parch`, `Pclass`) variables.  
- **Missing Values:**
  - `Age` → 177 missing entries (~20%)  
  - `Cabin` → 687 missing entries (~77%)  
  - `Embarked` → 2 missing entries  
- **Target Distribution:** Around **62% did not survive** and **38% survived**, showing a slight class imbalance.  
- **Average passenger age:** ~30 years; **Fare** varies widely, indicating strong skewness.  
- These findings guide our **next step — data cleaning**, where we’ll handle missing values and prepare the dataset for modeling.

---

<div align="center">
    
## 📌Step 3: Data Cleaning

</div>

## Objective
Handle missing values, remove irrelevant columns, and prepare the dataset for feature engineering.

### Key Tasks
- Impute missing values for **Age**, **Embarked**, and **Fare** if needed.  
- Decide what to do with **Cabin** (drop or extract deck info).  
- Drop or keep columns that are not useful for modeling (`PassengerId`, `Ticket`, `Name`).  
- Ensure the dataset is clean and ready for **feature engineering**.



In [4]:
# Fix missing values
train_df['Age'] = train_df['Age'].fillna(train_df['Age'].median())
train_df['Embarked'] = train_df['Embarked'].fillna(train_df['Embarked'].mode()[0])

# Drop columns safely (won't error if they are already missing)
train_df = train_df.drop(columns=['Cabin', 'Ticket', 'Name'], errors='ignore')

# Verify missing values after cleaning
print("\n--- Missing Values After Cleaning ---")
print(train_df.isnull().sum())

# Preview cleaned dataset
train_df.head()


--- Missing Values After Cleaning ---
PassengerId    0
Survived       0
Pclass         0
Sex            0
Age            0
SibSp          0
Parch          0
Fare           0
Embarked       0
dtype: int64


PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0            1         0       3    male  22.0      1      0   7.2500        S
1            2         1       1  female  38.0      1      0  71.2833        C
2            3         1       3  female  26.0      0      0   7.9250        S
3            4         1       1  female  35.0      1      0  53.1000        S
4            5         0       3    male  35.0      0      0   8.0500        S

### Step 3 Insights – Data Cleaning

- Missing values in `Age` and `Embarked` were filled with **median** and **mode** respectively.  
- `Cabin` dropped due to excessive missing values (~77%).  
- Irrelevant columns (`Ticket`, `Name`) were removed.  
- **No missing values remain**, and the dataset is now ready for feature engineering.

---

<div align="center">
    
## 📌Step 4: Feature Engineering

</div>

## Objective
Create new features from existing columns to improve the model’s predictive power and capture hidden patterns in the data.

### Key Tasks
- Create **FamilySize** = `SibSp + Parch + 1` to indicate family traveling together.  
- Create **IsAlone** = 1 if FamilySize == 1 else 0.  
- Extract **Title** from the passenger’s Name (e.g., Mr, Mrs, Miss, Master).  
- Optionally create **AgeGroup** (child, adult, senior) and **FareBand** to categorize fares.  
- Transform features to make them suitable for machine learning models.



In [5]:
# Step 4: Feature Engineering

# Family Size
train_df['FamilySize'] = train_df['SibSp'] + train_df['Parch'] + 1

# IsAlone
train_df['IsAlone'] = 0
train_df.loc[train_df['FamilySize'] == 1, 'IsAlone'] = 1

# AgeGroup
train_df['AgeGroup'] = pd.cut(train_df['Age'], bins=[0, 12, 18, 35, 60, 80],
                              labels=['Child','Teen','YoungAdult','Adult','Senior'])

# FareBand
train_df['FareBand'] = pd.qcut(train_df['Fare'], 4, labels=[1,2,3,4])

# Preview new features
train_df.head()


PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch     Fare  \
0            1         0       3    male  22.0      1      0   7.2500   
1            2         1       1  female  38.0      1      0  71.2833   
2            3         1       3  female  26.0      0      0   7.9250   
3            4         1       1  female  35.0      1      0  53.1000   
4            5         0       3    male  35.0      0      0   8.0500   

  Embarked  FamilySize  IsAlone    AgeGroup FareBand  
0        S           2        0  YoungAdult        1  
1        C           2        0       Adult        4  
2        S           1        1  YoungAdult        2  
3        S           2        0  YoungAdult        4  
4        S           1        1  YoungAdult        2

---

<div align="center">
    
## 📌Step 5: Encoding & Scaling

</div>

## Objective
Prepare the dataset for machine learning by converting categorical variables to numeric and scaling numeric features for better model performance.

### Key Tasks
- Convert **categorical variables** into numeric format:  
  - `Sex` → 0 for male, 1 for female  
  - `Embarked` → One-Hot Encoding (C, Q, S)  
  - `Title` → map to numeric codes if extracted earlier  
- Scale **numeric features** (like `Age` and `Fare`) if required for models sensitive to feature magnitude (e.g., Logistic Regression, SVM).  
- Ensure the dataset is fully numeric and ready for training.


In [6]:
# Step 5: Encoding & Scaling

from sklearn.preprocessing import StandardScaler

# 1️⃣ Encode Sex
train_df['Sex'] = train_df['Sex'].map({'male': 0, 'female': 1})

# 2️⃣ One-Hot Encode Embarked
train_df = pd.get_dummies(train_df, columns=['Embarked'], drop_first=True)

# 3️⃣ Encode AgeGroup and FareBand if created (optional)
if 'AgeGroup' in train_df.columns:
    train_df = pd.get_dummies(train_df, columns=['AgeGroup'], drop_first=True)

if 'FareBand' in train_df.columns:
    train_df['FareBand'] = train_df['FareBand'].astype(int)

# 4️⃣ Scale numeric features (Age, Fare)
scaler = StandardScaler()
train_df[['Age', 'Fare']] = scaler.fit_transform(train_df[['Age', 'Fare']])

# Preview the encoded & scaled dataset
train_df.head()


PassengerId  Survived  Pclass  Sex       Age  SibSp  Parch      Fare  \
0            1         0       3    0 -0.565736      1      0 -0.502445   
1            2         1       1    1  0.663861      1      0  0.786845   
2            3         1       3    1 -0.258337      0      0 -0.488854   
3            4         1       1    1  0.433312      1      0  0.420730   
4            5         0       3    0  0.433312      0      0 -0.486337   

   FamilySize  IsAlone  FareBand  Embarked_Q  Embarked_S  AgeGroup_Teen  \
0           2        0         1       False        True          False   
1           2        0         4       False       False          False   
2           1        1         2       False        True          False   
3           2        0         4       False        True          False   
4           1        1         2       False        True          False   

   AgeGroup_YoungAdult  AgeGroup_Adult  AgeGroup_Senior  
0                 True           False            False  
1                False            True            False  
2                 True           False            False  
3                 True           False            False  
4                 True           False            False

---

<div align="center">
    
## 📌Step 6: Model Training

</div>

## Objective
Train machine learning models to predict passenger survival using the cleaned, engineered, and encoded dataset.

### Key Tasks
- Split the dataset into **training** and **validation** sets to evaluate model performance.  
- Train multiple classification models:
  - Logistic Regression  
  - Decision Tree  
  - Random Forest  
  - Gradient Boosting / XGBoost (optional)  
- Compare performance on the validation set using metrics like **accuracy**, **precision**, **recall**, and **F1-score**.  
- Identify the **best performing model** for predictions on the test set.


In [8]:
# Step 6: Model Training

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report

# 1️⃣ Split features and target
X = train_df.drop('Survived', axis=1)
y = train_df['Survived']

# 2️⃣ Split into train and validation sets
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

# 3️⃣ Train Logistic Regression
log_model = LogisticRegression(max_iter=500)
log_model.fit(X_train, y_train)
y_pred_log = log_model.predict(X_val)
print("Logistic Regression Accuracy:", accuracy_score(y_val, y_pred_log))
print(classification_report(y_val, y_pred_log))

# 4️⃣ Train Decision Tree
dt_model = DecisionTreeClassifier(random_state=42)
dt_model.fit(X_train, y_train)
y_pred_dt = dt_model.predict(X_val)
print("Decision Tree Accuracy:", accuracy_score(y_val, y_pred_dt))
print(classification_report(y_val, y_pred_dt))

# 5️⃣ Train Random Forest
rf_model = RandomForestClassifier(n_estimators=100, random_state=42)
rf_model.fit(X_train, y_train)
y_pred_rf = rf_model.predict(X_val)
print("Random Forest Accuracy:", accuracy_score(y_val, y_pred_rf))
print(classification_report(y_val, y_pred_rf))


C:\ProgramData\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Logistic Regression Accuracy: 0.8044692737430168
              precision    recall  f1-score   support

           0       0.82      0.85      0.84       105
           1       0.77      0.74      0.76        74

    accuracy                           0.80       179
   macro avg       0.80      0.80      0.80       179
weighted avg       0.80      0.80      0.80       179

Decision Tree Accuracy: 0.7821229050279329
              precision    recall  f1-score   support

           0       0.80      0.84      0.82       105
           1       0.75      0.70      0.73        74

    accuracy                           0.78       179
   macro avg       0.78      0.77      0.77       179
weighted avg       0.78      0.78      0.78       179

Random Forest Accuracy: 0.8100558659217877
              precision    recall  f1-score   support

           0       0.83      0.86      0.84       105
           1       0.79      0.74      0.76        74

    accuracy                           0.81    

### Step 6 Insights – Model Training

- **Logistic Regression**
  - Accuracy: ~80%  
  - Precision & Recall: Slightly better for class 0 (did not survive) than class 1 (survived)  
  - F1-score: 0.84 (class 0), 0.76 (class 1)  

- **Decision Tree**
  - Accuracy: ~78%  
  - F1-score lower for class 1 (0.73) than class 0 (0.82)  
  - Model slightly underperforms compared to Logistic Regression and Random Forest  

- **Random Forest**
  - Accuracy: ~81% (highest among tested models)  
  - Balanced performance across classes  
  - F1-score: 0.84 (class 0), 0.76 (class 1)  
  - Appears to generalize better than single Decision Tree  

### 🔑 Key Insights
- Ensemble methods like **Random Forest** perform slightly better than single models.  
- All models show **higher performance for class 0** (passengers who did not survive) due to slight class imbalance.  
- Logistic Regression is simple and performs reasonably well, but Random Forest is the **best choice** for final predictions.
